EIPS/eip-3171.md

```diff
@@ -0,0 +1,148 @@
+not involve the init code hash. It is otherwise identical to `CREATE2` opcode.
+
+This opcode allows users to create accounts at an address independent of the bytecode deployed to the address.
+
+## Abstract
+
+TBD
+
+## Motivation
+
+The core motivations of the `CREATE3` opcode are the same as [EIP-1014](./eip-1014.md), i.e. deploying a smart contract at a
+deterministic address.
+
+This variation removes the hash of the init code from the preimage. This enables the address of contract created via `CREATE3`
+to be computed on-chain knowing only the deployer address and salt.
+
+It is possible to compute the address of a contract created with `CREATE2` without making any external calls. For example,
+Uniswap V2 [uses this](https://github.com/Uniswap/uniswap-v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol#L18)
+to save gas in computing the pair address from the Uniswap V2 Router. The gas cost of the `pairFor` method which
+computes the pair address for a given pair of tokens is only ~649 gas, while an external call + an SLOAD from mapping
+may cost upwards of 3k gas. A swap that is routed through multiple pairs requires multiple calls of this method.
+
+This is feasible only because the Uniswap V2 Pair does not have any constructor arguments, which means the init code hash
+can be hard coded in the library. However, the init code hash cannot be hard coded when the deployed smart contract has
+constructor arguments. In order to compute the init code hash, one must append the constructor arguments to the init code
+before computing the init code hash. Thus, in order to compute the address of an account created via `CREATE2`, it is 
+necessary to include the entire init code without the constructor arguments, and hash 
+`ceil(len(init_code_base + constructor_arguments) / 32)` words on-chain.
+
+Later versions of solidity introduced immutables, which allow for further gas savings but must be initialized in the 
+constructor.
+
+## Specification
+
+Adds a new opcode (`CREATE3`) at `0xf6`, which takes 4 stack arguments: endowment, memory_start, memory_length, salt. 
+Behaves identically to `CREATE2` (`0xf5`), except using `keccak256( 0xfe ++ address ++ salt )[12:]`,
+which does not include the hash of the init code and uses a different byte prefix.
+
+The `CREATE3` opcode has the same `gas` schema as `CREATE2`, but with one fewer `GSHA3WORD` opcode, to account for the difference
+in lengths of the preimages (`53` bytes instead of `85` in `CREATE2`).
+The `hashcost` is deducted at the same time as memory-expansion gas and `CreateGas` is deducted: _before_ evaluation
+of the resulting address and the execution of `init_code`.
+
+- `0xfe` is a single byte, 
+- `address` is always `20` bytes, 
+- `salt` is always `32` bytes (a stack item). 
+
+The preimage for the final hashing round is thus always exactly `53` bytes long (does not include hash of the init code).
+
+## Rationale
+
+### Address formula
```

Comment:
I'm don't think it is required to use `0xfe` instead of `0xff`. Every element of the `create2` and `create3` formula is fixed, so it's not trivial to create a collision. I recommend maintaining the `0xff` beginning over introducing a new one.

Reply:
I agree, changed to `0xff` and the collisions are prevented by the difference in preimage length